pkg/validation/errors.go

```diff
@@ -0,0 +1,36 @@
+package validation
```

Comment:
> That allows package consumers to implement that however they wont, as opposed to be restricted to use the helpers.

Unexporting the interface doesn't prohibit another package from satisfying it. The interface can be documented and still implemented by custom errors (this is perhaps a not so obvious benefit of duck typing).

```go
package validation

type remediable interface {
	Remediation() string
}

func IsRemediable(err error) bool {
	_, ok := err.(remediable)
	return ok
}
```

```go
package foo

// This works just fine with IsRemediable()
type myRemediableError struct {}
func (myRemediableError) Remediation() string { ... }
func (myRemediableError) Error() string { ... }
```

Regarding exposure of the interface, that gives an opportunity for consumers to return it which should be discouraged in favor of opaque errors (rule of thumb). That also mean any utility funcs like `NewRemediableErr` should return `error`.

The utility functions are just that. You can ignore them and define your own remediable error, or use them for convenience in error handling.

Dave Cheney does a great job explaining this kind of error handling: https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully.

> It allows to check if the error is remediable and get the message in one call. As opposed to two, which implies checking for the interface implementation twice.

My motivation is readability at the consumer site and my feeling was the second form below is more readable than the first - I'm OK either way, though I'm not sure there's a problem with a double interface check.

```go
if rerr, ok := validation.ErrorRemediation(err); ok {
	handleRemediation(rerr)
}
```
vs
```go
if validation.IsRemediable(err) {
  handleRemediation(validation.Remediation(err))
}
```

Reply:
> Unexporting the interface doesn't prohibit another package from satisfying it. The interface can be documented and still implemented by custom errors (this is perhaps a not so obvious benefit of duck typing).

Sure you can implement the unexported interface, but you need to know about it. In that case, why wouldn't you use the code as documentation? Plus documenting it but not exposing in the package's API creates a hidden dependency. If we were to change our `Remediable` interface, the code implementing it could break silently.

> Regarding exposure of the interface, that gives an opportunity for consumers to return it which should be discouraged in favor of opaque errors (rule of thumb).

You mention trying to avoid other code from returning it in favor of opaque errors. Isn't that going a bit too far? Even if we didn't export it, other code can define its own copy of the interface and return it. I'm in favor of using the type system to ensure correctness when possible, but trying to enforce a particular error pattern seems too much.

> That also mean any utility funcs like NewRemediableErr should return error.

I don't think I agree here. I buy the argument for the code that actually generates the error since it helps decouple that package from the caller. But this is just a constructor for the error. Callers might actually want *this* error.

> I'm OK either way, though I'm not sure there's a problem with a double interface check.

Performance was my concern. But let's go for legibility and we can come back to it if it becomes a problem.

Reply:
Ummm I just started playing around with the code.

Now I realize that in order to return remediable errors as an interface from the constructors, I need to embed `error` into `Remediable` or else the returned value doesn't directly implement `error`.

I don't like that, it pollutes the `Fixable` interface just to be able to use the returned value as an `error` elsewhere. I'm gonna change it to return `error` as you suggest, sorry for the back and forth.

Reply:
> Sure you can implement the unexported interface, but you need to know about it. In that case, why wouldn't you use the code as documentation? Plus documenting it but not exposing in the package's API creates a hidden dependency. If we were to change our Remediable interface, the code implementing it could break silently.

FWIW, its a reasonably widely accepted pattern and the standard library also buys into it: [`errors.Is()`](https://cs.opensource.google/go/go/+/refs/tags/go1.20.1:src/errors/wrap.go;l=53), [`errors.As()`](https://cs.opensource.google/go/go/+/refs/tags/go1.20.1:src/errors/wrap.go;l=113) and [`errors.Unwrap()`](https://cs.opensource.google/go/go/+/refs/tags/go1.20.1:src/errors/wrap.go;l=16) all document their expectations without any explicit interface.

> If we were to change our Remediable interface, the code implementing it could break silently.

Yes, absolutely. The interface is small and simple though, so the risk seems low.